designs/2020-generic-ast-support/README.md

```diff
@@ -0,0 +1,57 @@
+- Repo: eslint/eslint
+- Start Date: 2020-06-04
+- RFC PR: 
+- Authors: Ilya Volodin
+
+# Generic AST support
+
+## Summary
+
+Allow ESLint to traverse and lint generic (non estree compliant AST) and allow parsers to specify information about AST that would drive ESLint's traversal and features.
+
+## Motivation
+
+ESLint has cemented itself as a go to tool for JavaScript/TypeScript ecosystem for style checking and linting. However, there are a lot of connected languages/technologies that are used every day by ESLint users that are either lacking linters, or require yet another tool to be setup and configured for the repository. Examples include HTML, CSS, SCSS, LESS, GraphQL, JSON, etc.
+With addition of config overrides users can now setup separate configuration per glob, so that would allow them to have a single instance of ESLint and a single configuration that would allow linting of the entire repository, including all non javascript/typescript files.
+
+## Detailed Design
+
+Currently ESLint has a lot of hardcoded assumptions about AST. For example, it assumes that nodes are identified by the property called `type` and that property exists on every node. It assumes that top-most node is always `Program` and that AST supports code-path-analysis (which is very specific to JavaScript). This RFC proposes to modify expected output from the `parseForESLint` method of the custom parsers to include additional property called `parserSupportOptions` of type object with four new properties (for now):
+ - nodeIdentifier: string - Specifies name of the ASTNode property that identifies node. Defaults to `type` for backwards compatibility.
+ - supportCodePathAnalysis: Boolean - Indicates AST's support for built-in CodePathAnalysis support. Defaults to `true` for backwards compatibility.
```

Comment:
[Like @kaicataldo](https://github.com/eslint/rfcs/pull/56#issuecomment-640092811), I could see us adding more thorough support for arbitrary ASTs in the future including full support for scope and code path analysis. If we do that, we'd need to provide a way for parsers to supply their own code path analysis. In this hypothetical future, I imagine we'd standardize the existing code path analysis interface or something very similar to it as a public API and allow parsers to pass in an alternative implementation matching that interface. That could make this boolean property redundant or lead to invalid combinations of the two values (i.e. `supportsCodePathAnalysis: false, codePathAnalysis: {Implementation}` makes no sense).

Like you, I don't want to try to standardize pluggable code path analysis in this RFC, so my only goal is to avoid something that we'll soon want to make redundant. Instead of this being a boolean property, what if it's called `codePathAnalyzer`, and for now its only possible value is `null`? If the property is absent, we keep the existing behavior and use the built-in code path analyzer, but if it's present and `null`, we disable code path analysis. This allows us a built-in hook for whenever we standardize a public API for pluggable code path analyzers.

We could also move this property to be at the same level as `scopeManager`.

Reply:
I would be fine with this change, but it can be pretty confusing to somebody from outside looking at the API. If you pass in `null` it all-of-a-sudden disables a feature. Seems a bit magical. What I would prefer instead is to have a variable `codePathAnalyzer` which would right now be of type `boolean` (defaulting to `true`). In the future we can make it accept function or boolean. Boolean `true` would mean enable codePathAnalyzer, `false` - disable, `function` means use this for codePathAnalyzer.

Reply:
You're right, that would be pretty confusing. And from the implementation side, handling `codePathAnalyzer` of type `boolean | function` isn't difficult at all.